### DISCUSSION TITLE
Create a channel based on a frontend user regisration

### DISCUSSION THREAD

**Discussion by dolphiq-allard** [craftcms/cms#12620] [Q&A]
I am currently working on a project where I'd like frontend users to be able to register an account. This works like a charm using the Craft CMS tools out of the box.

However: based on a specific (custom) field in this registration e.g. `Company Display name` I'd like a channel to be automatically created with access only to the registered user using the custom field. So result should be a Channel named e.g. `Pixel and Tonic` in which the user can add entries which can be shared on the internet via `https://example.com/pixel-and-tonic/example-entry`

Goal is to allow the user to allow CRUD on entries in this specific channel without being able to access/edit other users channels/entries.

What would be the best way to achieve this?

Cheers

**Comment by brandonkelly** [comment]
I wouldn’t recommend this approach, as the UI was not designed for a very large number of sections. And sections are expected to only be created locally, as they are included in your project config, which should never change on production.

You could give users a single entry within a single section. If they are the author you can use the existing author permissions to restrict their access to only save their own entry. Alternatively you could do your own permission enforcement, by hooking into the `craft\services\Elements::EVENT_AUTHORIZE_SAVE` event.

```php
use craft\elements\Entry;
use craft\events\AuthorizationCheckEvent;
use craft\services\Elements;
use yii\base\Event;

Event::on(
    Elements::class,
    Elements::EVENT_AUTHORIZE_SAVE,
    function(AuthorizationCheckEvent $event) {
        if (
            $event->element instanceof Entry &&
            $event->element->getSection()->handle === 'userSectionHandle' &&
            $event->element->someUserIdField == $event->user->id
        ) {
            $event->authorized = true;
        }
    }
);
```

**Reply by dolphiq-allard** [reply]
Completely looked over the fact that channels are stored in the settings, so that combined with performance makes that approach useless.

A dedicated entry (with child entries) per user would work for my case. However since they can register in the app publicly, the entry should be automatically created and have tje correct rights for only the registered user. 

Is this programatically possible.

for context, the goal is to create an app that allows users to publicly register an account, create an entry which can serve as a public profile page (using entry type: "Profile page") and then create entries (using entry type: "Article") within that entry which can contain blocks of information, files and images which are not indexed by search engines and can be shares with one or more invitees, ideally optionally protected with a password.

Desired structure should look like this:

```
Users (section)
├--- User 1 (Entry: Profile Page, public, e.g. https://example.com/pixel-and-tonic)
│       ├--- Article 1 (Entry: Article Page, noindex, possible password, e.g. https://example.com/pixel-and-tonic/article-1)
│       ├--- Article 2 (Entry: Article Page, noindex, possible password)
│       ├--- Article 3 (Entry: Article Page, noindex, possible password)
│       ├--- Article 4 (Entry: Article Page, noindex, possible password)
├--- User 2 (Entry: Profile Page, public)
│       ├--- Article 1 (Entry: Article Page, noindex, possible password)
│       ├--- Article 2 (Entry: Article Page, noindex, possible password)
...
```

I am trying to figure out if such a thing a) can be done in Craft, which would be awesome since the base functionality is rock solid and b) if this would work performance wise. Or should such a portal be build using Laravel/Livewire/BackPack?

We now use Craft CMS mostly as a content management system, however I think there's way more potential for us and our clients in the software for portal-esque solutions.

Any advice is highly appreciated.

**Reply by brandonkelly** [reply]
For profiles, you might be able to just get away with just putting all the profile data directly on the user accounts with custom fields. No additional entry needed.

If you do have a good reason to need to create a dedicated entry per user (e.g. you need the profiles to be translatable), sure it’s possible. Give the entry type a Users relation field for tracking who the entry is for, and then create the entries when users are first saved:

```php
use craft\elements\Entry;
use craft\elements\User;
use craft\events\ModelEvent;
use yii\base\Event;

Event::on(User::class, User::EVENT_AFTER_SAVE, function(ModelEvent $event) {
    if ($event->isNew) {
        /** @var User $user */
        $user = $event->sender;
        $section = Craft::$app->getSections()->getSectionByHandle('userProfiles');
        $entry = new Entry([
            'sectionId' => $section->id,
            'typeId' => $section->getEntryTypes()[0]->id,
            'myUsersField' => [$user->id],
            // ...
        ]);
        Craft::$app->elements->saveElement($entry);
    }
});
```

Or you could create the entry when users are first activated instead, via `craft\services\Users::EVENT_AFTER_ACTIVATE_USER` – up to you.

Rather than store the article entries nested below the main user profiles, it would probably be better to just give them their own separate section. Again, with a Users field used to relate the articles to the user accounts.

**Reply by dolphiq-allard** [reply]
Thanks again. The reason I'd like users to be able to create entries is that they should be able to create an "article" consisting of a few content blocks which they can share with their clients for approval. Obviously these articles, or semi private webpages can only be edited by the user who created the page. Furthermore it should be possible to have a low level of styling on those "client" pages, such as colors, maybe typography and a logo/images.

Since I don't know how many frontend user will register and how many client proposals they will each add, it should be automated on registation with either creating a section and attaching rights the the section based on the user, or entries.

Does this make sense?

**Reply by brandonkelly** [reply]
My understanding was that articles would be separate entries though. So those should definitely be a channel. But what about the top-level profile data?

**Reply by dolphiq-allard** [reply]
Thanks for your reply. What I did now is create a structure with the first level being the corporate profile page and the second level being the actual page holding the content. Furthermore as per your suggestion I use the front end user to set some top-level data.

So far it looks like this will work just right with some custom controllers to manage saving the data across both users and entries.
